backstory/src/test/java/com/google/sps/vision/VisionManagerImplTest.java

```diff
@@ -0,0 +1,218 @@
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+
+    try {
+      BufferedImage bImage = ImageIO.read(new File(reference));
+      ImageIO.write(bImage, fileType, bos);
+    } catch (IOException exception){
+      System.err.println("IOException while reading image");
+    }
+
+    return bos.toByteArray();
+  }
+
+  /** Detects labels in the image specified by the image byte data by calling the Vision API. */
+  private static List<EntityAnnotation> detectLabelsFromImageBytes(byte[] bytes) throws IOException {
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    List<EntityAnnotation> labels;
+
+    Image img = Image.newBuilder().setContent(ByteString.copyFrom(bytes)).build();
+    Feature feat = Feature.newBuilder().setType(Feature.Type.LABEL_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feat).setImage(img).build();
+    requests.add(request);
+
+    // Initialize client that will be used to send requests. This client only needs to be created
+    // once, and can be reused for multiple requests. After completing all of the requests the
+    // client will be automatically closed, as it is called within the try.
+    try (ImageAnnotatorClient client = ImageAnnotatorClient.create()) {
+      BatchAnnotateImagesResponse response = client.batchAnnotateImages(requests);
+      List<AnnotateImageResponse> responses = response.getResponsesList();
+
+      // There is only one image in the batch response (only supports uploading one image at a time
+      // right now)
+      AnnotateImageResponse res = responses.get(0);
+
+      if (res.hasError()) {
+        System.out.format("Error: %s%n", res.getError().getMessage());
+        return new ArrayList<EntityAnnotation>();
+      }
+
+      // For full list of available annotations, see http://g.co/cloud/vision/docs
+      labels = res.getLabelAnnotationsList();
+    }
+
+    return labels;
+  }
+
+  /**
+   * Tests paritions on public VisionManagerImpl(byte[] rawImageData);
+   *  rawImageData is null.
+   *  Expects IllegalArgumentException to be thrown.
+  */
+  @Test(expected = IllegalArgumentException.class)
+  public void constructorNullImageData() throws IOException {
+    VisionManager visionManagerNullImage = new VisionManagerImpl(nullRawImageData);
+  }
+
+  /**
+   * Tests paritions on VisionManagerImpl(byte[] rawImageData);
+   *  rawImageData is empty.
+   *  Expects IllegalArgumentException to be thrown.
+  */
+  @Test(expected = IllegalArgumentException.class)
+  public void constructorEmptyImageData() throws IOException {
+    VisionManager visionManagerNullImage = new VisionManagerImpl(emptyRawImageData);
+  }
+
+  /**
+   * Tests paritions on VisionManagerImpl(byte[] rawImageData);
+   *  rawImageData is non-empty image data.
+  */
+  @Test
+  public void constructorRawImageData() {
+    List<EntityAnnotation> labelAnnotations = new ArrayList<>();;
```

Comment:
You declare this list more than once in methods. You could probably set this up in the ```@Before``` tag.

Reply:
I did that - thanks for the tip.

Reply:
Great!